x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APMTracer.java

```diff
@@ -102,57 +139,77 @@ protected void doStart() {
                 .build()
         );
 
-        openTelemetry = OpenTelemetrySdk.builder()
+        var openTelemetry = OpenTelemetrySdk.builder()
             .setTracerProvider(provider)
             .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
             .build();
-        tracer = openTelemetry.getTracer("elasticsearch", Version.CURRENT.toString());
+        var tracer = openTelemetry.getTracer("elasticsearch", Version.CURRENT.toString());
+
+        this.services = new APMServices(provider, tracer, openTelemetry);
     }
 
     @Override
     protected void doStop() {
-        final SdkTracerProvider provider = this.provider;
-        if (provider != null) {
-            provider.forceFlush().join(10L, TimeUnit.SECONDS);
-            provider.shutdown().join(30L, TimeUnit.SECONDS);
-        }
+        doStop(true);
     }
 
     @Override
-    protected void doClose() {}
+    protected void doClose() {
+
+    }
+
+    private void doStop(boolean wait) {
+        var services = this.services;
+        this.services = null;
+        if (services == null) {
+            return;
+        }
+        this.spans.clear();// discard in-flight spans
+        shutdownPermits.tryAcquire();
+        services.provider.shutdown().whenComplete(shutdownPermits::release);
+        if (wait) {
```

Comment:
Could we just move this block into the `doStop()` method? Seems odd to pass in a flag when we could just write the code.